### PR TITLE
#160451508 User should be able to tag articles

### PR DESCRIPTION
**Pull request by kampkelly**
#### What does this PR do?
This pull request adds tagging of articles.

#### Description of Task to be completed?
- Write tests for tagging articles
- Write the code to tag an article

#### How should this be manually tested?
After cloning the project, run `npm install`, then `npm run watch` to start the server. Using postman and as a signed in user, make requests to the following api endpoints.

To create a tag: /api/v1/tags -> POST request (params: name)
To retrieve tags: /api/v1/tags -> GET request
To tag articles: /api/v1/articles -> POST request (params: title, body, description, tags)

#### Any background context you want to provide?
None
#### What are the relevant pivotal tracker stories?
#160451508
#### Questions:```
None

### CODE REVIEW

File: server/controllers/articleController.js [hunk 1]
```diff
@@ -17,9 +18,21 @@ export default class ArticleController {
   static createArticle(req, res) {
     const { title, body, description } = req.body;
     const imgUrl = (req.file ? req.file.secure_url : '');
-    Article.create({
-      title, body, userId: req.user.id, description, slug: `${slug(title)}-${uuid()}`, imgUrl
-    }).then(article => res.status(201).json({ message: 'article created successfully', success: 'true', article }))
+    getTagIds(req, res).then((tagIds) => {
+      Article.create({
```
Comment (kampkelly) [inline server/controllers/articleController.js:28]: This change here gets the ids of the tags sent along with the request. Then the article is created and the tagged. Then, the article is returned with the tags.

File: server/controllers/articleController.js [hunk 2]
```diff
@@ -91,15 +104,27 @@ export default class ArticleController {
...
-      }
+    getTagIds(req, res).then((tagIds) => {
+      Article.findOne({
+        where: { slug: req.params.slug, userId: req.user.id },
+      }).then((article) => {
+        if (article === null) {
+          res.status(404).json({ message: 'article does not exist', success: false });
+        } else {
+          article.update(req.body)
+            .then((updatedArticle) => {
```
Comment (kampkelly) [inline server/controllers/articleController.js:132]: This change here gets the ids of the tags sent along with the request. Then the article is found, updated and tagged. Then, the article is returned with the tags.

File: server/controllers/tagController.js [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  * @description CRUD operations on Article
+  */
+export default class TagController {
+  /**
+  * @description -This method creates tag for an authenticated user
+  * @param {object} req - The request payload sent from the router
+  * @param {object} res - The response payload sent back from the controller
+  * @returns {object} - status, message and tag detail
+  */
+  static createTag(req, res) {
```
Comment (kampkelly) [inline server/controllers/tagController.js:15]: The createTag gets the array of tags from the body and creates it if it doesn't exist

File: server/helpers/tags/getTagIds.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+      const collectTagIds = tag => tag.id;
+      const tagIds = tags.map(collectTagIds);
+      resolve(tagIds);
+    })
+      .catch(error => res.status(500).json(error));
+  });
+  return promise;
+};
+
+export default getTagIds;
```
Comment (kampkelly) [inline server/helpers/tags/getTagIds.js:26]: This helper returns the ids of the tags sent in the request to the create and update articles endpoints.

File: server/helpers/validations/joiValidations.js [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+      // this regex replaces the string (\") that is returned with the json response
+      const getErrorMessages = error => error.message.replace(/"/g, '');
+      res.status(422).json({ message: err.details.map(getErrorMessages), status: 'failed', });
+    } else if (!err) {
+      return next();
+    }
+  });
+};
+
+export default joiValidations;
```
Comment (kampkelly) [inline server/helpers/validations/joiValidations.js]: This block of code is used by all the validations in the middlewares. So it has been moved to it s own file.

**Review by onwuvic — COMMENTED** [review]
LGTM

**Comment by onwuvic** [comment]
Resolve your merge conflict

File: .eslintignore [hunk 6]
```diff
@@ -1,2 +1,3 @@
 /server/models/*.js
 /server/migrations/*.js
+/server/seeders/*.js
```
Comment (s-akinrele) [inline .eslintignore]: add a new line here

File: server/controllers/articleController.js [hunk 7]
```diff
@@ -68,15 +82,15 @@ export default class ArticleController {
...
-      include: [{
-        model: Like,
-        as: 'likes'
-      }]
+      include: [
+        { model: Tag, as: 'Tags', through: 'ArticlesTags' },
+        { model: Like, as: 'likes' }
+      ]
     }).then((article) => {
       if (article === null) {
```
Comment (s-akinrele) [inline server/controllers/articleController.js]: change as discussed.

File: server/controllers/articleController.js [hunk 8]
```diff
@@ -111,15 +138,20 @@ export default class ArticleController {
...
   */
   static deleteArticle(req, res) {
-    Article.destroy({
-      where: { slug: req.params.slug, userId: req.user.id }
+    Article.findOne({
+      where: { slug: req.params.slug, userId: req.user.id },
     }).then((article) => {
-      if (article === 0) {
-        res.status(404).json({ message: 'article does not exist', success: 'false' });
+      if (article === null) {
```
Comment (s-akinrele) [inline server/controllers/articleController.js]: fix as discussed.

File: server/helpers/validations/joiValidations.js (hunk 5 above)
Comment (s-akinrele) [reply server/helpers/validations/joiValidations.js]: rename this method

**Review by s-akinrele — CHANGES_REQUESTED** [review]
Fix changes as discussed

File: server/controllers/articleController.js [hunk 9]
```diff
@@ -72,17 +87,22 @@ export default class ArticleController {
...
-      include: [{
-        model: ArticleComment,
-        include: [{
-          model: User, attributes: ['username', 'email'],
-        }],
-      }, {
-        model: Like,
-        as: 'likes'
-      }]
+      // include: [
```
Comment (s-akinrele) [inline server/controllers/articleController.js]: remove commented code.

**Review by s-akinrele — APPROVED** [review]

**Pull request merged**